### PR TITLE
feat: memote

### PR DESCRIPTION
**Pull request by mihai-sysbio**
### Main improvements in this PR:
As a follow-up to #162, this PR merges in the `devel` and `feat/yaml-validation` branches in order to convert the Travis CI pipeline that runs memote to GH Actions. Just as before, the [`history_report.html`](https://mihai-sysbio.github.io/yeast-GEM/history_report.html) gets commited to the `gh-pages` branch automatically.

### CODE REVIEW

**Comment by Midnighter** [comment]
Some thoughts on the GitHub action:

* Rather than using a number of conditions you could set up two different actions. One that runs on `push` and another that runs on `pull_request`.
* You may want to use `memote run --skip-unchanged` in more places.

**Comment by Midnighter** [comment]
> @Midnighter can we directly take results from results.db for this, or do we have to re-run the memote suit for those 2 models?

At the moment, you would have to re-run memote. However, I have been wanting this feature for a very long time... Let me see what I can do.

**Comment by mihai-sysbio** [comment]
Items one and two are implemented. In addition to the biomass rxn issue, memote doesn't like the uncommitted model
> [critical: Please git commit or git stash all changes before running the memote suite.](https://github.com/mihai-sysbio/yeast-GEM/runs/881789519?check_suite_focus=true#step:8:9)

It sounds to me like the model conversion should be commented out for now, and enabled when the two issues above are fixed.

**Comment by mihai-sysbio** [comment]
@BenjaSanchez as you mentioned above, I agree with elevating what's left from these issues and cannot be solved now back to #162

File: .github/workflows/memote-history.yml [hunk 1]
```diff
@@ -35,11 +35,18 @@ jobs:
         ref: ${{ steps.setup.outputs.deployment }}
         path: ${{ steps.setup.outputs.other-repo }}
 
+    - name: Convert model
+      run: |
```
Comment (Midnighter) [inline .github/workflows/memote-history.yml:39]: You can consider the following so you don't need to escape line breaks.
```suggestion
      run: >
```

**Comment by Midnighter** [comment]
> Items one and two are implemented. In addition to the biomass rxn issue, memote doesn't like the uncommitted model
> 
> > [critical: Please git commit or git stash all changes before running the memote suite.](https://github.com/mihai-sysbio/yeast-GEM/runs/881789519?check_suite_focus=true#step:8:9)
> 
> It sounds to me like the model conversion should be commented out for now, and enabled when the two issues above are fixed.

If you are not interested in that file, you could indeed just `git stash` or remove it.

File: .github/workflows/memote-history.yml (hunk 1 above)
Comment (mihai-sysbio) [reply .github/workflows/memote-history.yml:39]: Good suggestion! I would rather keep the line breaks for easier copy-pasting into different shell, although I think `zsh` wouldn't need these. @BenjaSanchez  any preference for `|` or `>`?

**Comment by mihai-sysbio** [comment]
@Midnighter stashing would defeat the point raised above by @BenjaSanchez through:
```python
model = io.read_yeast_model(make_bigg_compliant=True)  # load with BiGG ids
io.write_yeast_model(model)
```

File: .github/workflows/memote-history.yml (hunk 1 above)
Comment (BenjaSanchez) [reply .github/workflows/memote-history.yml:39]: I'm afraid I don't understand the implications of either, could you ellaborate?

Comment (Midnighter) [reply .github/workflows/memote-history.yml:39]: Sorry, I was not very clear. `: >` means you can break the lines as you please and each newline will be transformed into a space. So you can break up a command over multiple lines without escaping them. `: |` on the other hand considers every line separately. This is useful for separate commands or you need to escape line breaks as you have done here with `\`.

You can read more about it in the spec https://yaml.org/spec/1.2/spec.html#id2760844

**Comment by Midnighter** [comment]
> @Midnighter stashing would defeat the point raised above by @BenjaSanchez through:
> 
> ```python
> model = io.read_yeast_model(make_bigg_compliant=True)  # load with BiGG ids
> io.write_yeast_model(model)
> ```

I see two immediate approaches. The problem is that you want a model with BiGG compliant IDs that improves the behaviour on some memote tests. Memote complains about a dirty repository because we did not want to be responsible for committing any other files than the results. We could make memote smarter about this but this is the current behaviour. Possible solutions:

1. You create a BiGG compatible model file as part of your pre-commit hooks just like YAML and JSON and point memote at that file.
2. You could also call the memote test suite from within Python. That means, you would load the model file in a BiGG compliant model and then directly pass the model instance to the memote function call.

I prefer 1. to 2. because it is much more reproducible. If a user downloads your normal XML model in scenario 2., they will get a different score and be confused.

**Comment by mihai-sysbio** [comment]
@Midnighter cool ideas, both of them. Thanks for the background story. Some thoughts:
1. I have no idea if this would lead to the versioning of two `xml` files - there must be a reason why BiGG complaint model is not the default one.
2. Do you mean it's possible to run `memote report history` from within Python?

**Comment by Midnighter** [comment]
> Do you mean it's possible to run memote report history from within Python?

Certainly, you can look at the CLI commands in the `cli.py` module which then call functions in the https://github.com/opencobra/memote/blob/develop/src/memote/suite/api.py module. In this case, I meant [test_model](https://github.com/opencobra/memote/blob/4c6739eb1b6ba2c50d3db905347fa0b880373bca/src/memote/suite/api.py#L67) which is the equivalent of `memote run`.

File: .github/workflows/memote-history.yml (hunk 1 above)
Comment (BenjaSanchez) [reply .github/workflows/memote-history.yml:39]: @Midnighter thanks for the explanation. @mihai-sysbio `>` indeed seems simpler so I would use that one

**Comment by BenjaSanchez** [comment]
> I have no idea if this would lead to the versioning of two `xml` files

It would indeed, and this is something I think should be avoided: having 2 different `xml` would create confusion for users + make it harder for developers, as 2 `xml` would have to be updated on each commit, one using cobratoolbox and one using cobrapy.

> there must be a reason why the BiGG complaint model is not the default one

BiGG annotation was only added recently (#188), so the model has existed for more than 10 years with the current ids. We *could* change the ids and use the [BiGG dictionaries](https://github.com/SysBioChalmers/yeast-GEM/tree/master/ComplementaryData/databases) to convert when loading into Matlab using the `loadYeastModel` wrapper, but that would not work for users that directly download the model instead of cloning the repo / using the wrappers. Therefore, I think keeping the current ids is the most sensible option in terms of guaranteeing backward compatibility to all users. I'm ofc up for discussing it further, but maybe in a separate issue?

WRT the 2 possibilities @Midnighter suggests then, I would say the 2nd one would be the way to go for now; we can then ease any worried users with a disclaimer in the README/landing page mentioning the extra steps taken.

Finally, note that met/rxn ids is not the only thing that changes when the model is taken through an I/O cycle in cobrapy, but will be once https://github.com/SysBioChalmers/yeast-GEM/issues/218 is fully addressed (although that will likely take time)

**Comment by mihai-sysbio** [comment]
I'm trying the 2nd suggestion, but I get stuck.
```
import memote
import git

repoRM = memote.RepoResultManager(git.Repo(), "sqlite:///results.db")
historyM = memote.HistoryManager(repoRM, 'devel')
print(historyM._history)
memote.history_report(historyM)
```

History is `None`, and the last line fails.

Now, even if it would succeed, [`history_report`](https://github.com/opencobra/memote/blob/4c6739eb1b6ba2c50d3db905347fa0b880373bca/src/memote/suite/api.py#L153) does not run like [`test_model`](https://github.com/opencobra/memote/blob/4c6739eb1b6ba2c50d3db905347fa0b880373bca/src/memote/suite/api.py#L67) - it cannot take the model as a parameter. This means it is likely that we would get to the same error as [above](https://github.com/SysBioChalmers/yeast-GEM/pull/239#issuecomment-660082847), where the HistoryManager dislikes the uncommitted model. I still can't get this 2nd suggestion to run - any ideas?

**Comment by Midnighter** [comment]
Let me see, I think that `.github/workflows/memote-run.yml` should be straight forward since it only needs to run `test_model`, right?

Replicating `.github/workflows/memote-history.yml` may indeed be trickier.

1. You first test the model but this time memote actually commits the result (to the deployment branch). You should be able to see how that's done in the `cli.py` module. And I'd suggest that you read the relevant information from the configuration file, i.e., do not hard code values.
2. After that it should be possible to simply run `memote report history` since you don't need access to the model at that point.

**Comment by BenjaSanchez** [comment]
update: the biomass issue https://github.com/opencobra/memote/issues/689 is now solved: using the latest version of memote 0.11.1 (which we can get in this branch if we `pip-compile --upgrade ci-requirements.in`) only 1 biomass reaction is detected. Thanks @Midnighter for looking into it :)

PS @mihai-sysbio I noticed you combined both the memote badge + Travis badge into a single one. I see value in having both (one directs to the history report and the other to the test suite to see if it's passing), but we can get back to it once the history report is properly created.

**Comment by Midnighter** [comment]
I was thinking that I should add a GH workflow to memote in addition to or instead of the Travis CI one. I'm glad that you're already going ahead with that.

**Comment by mihai-sysbio** [comment]
@Midnighter that would be great.

This very large PR has been in progress for a long time, but finally it passes all checks. It would be nice to merge it and let  subsequent ideas/issues surface as different PRs.
Note: the `fix/requirements` PR #256 should be merged before this one.

**Pull request merged**